Change Github wrong link to helm Artifact hub (#1032)
redpanda-data/documentation: modules/manage/partials/iceberg/query-iceberg-topics.adoc

endif::[]

ifdef::env-byoc[]
Depending on the processing engine and your Iceberg catalog implementation, you may also need to create a table to point your data lakehouse to the table location in the catalog.

For BYOC clusters on AWS and GCP, the bucket name and table location are as follows:

|===
| Bucket name | Iceberg table location

| `redpanda-cloud-storage-<cluster-id>`
| `redpanda-iceberg-catalog/redpanda/<topic-name>`

|===
endif::[]

If your engine needs the full JSON metadata path, use the following:

```
redpanda-iceberg-catalog/metadata/redpanda/<topic-name>/v<version-number>.metadata.json
```

This provides read access to all snapshots written as of the specified table version (denoted by `version-number`).

NOTE: Redpanda automatically removes expired snapshots on a periodic basis. Snapshot expiry helps maintain a smaller metadata size and reduces the window available for <<time-travel-queries,time travel>>.

== Query examples

ifndef::env-byoc[]

> To follow along with the examples on this page, suppose you produce the same stream of events to a topic `ClickEvent`, which uses a schema, and another topic `ClickEvent_key_value`, which uses the key-value mode. The topics have xref:manage:tiered-storage.adoc[Tiered Storage] configured to an AWS S3 bucket. A sample record contains the following data:

Reviewer: suggest making Tiered Storage a glossterm instead of a link

endif::[]

ifdef::env-byoc[]
To follow along with the examples on this page, suppose you produce the same stream of events to a topic `ClickEvent`, which uses a schema, and another topic `ClickEvent_key_value`, which uses the key-value mode. The topic's Iceberg data is stored in an AWS S3 bucket. A sample record contains the following data:
endif::[]

[,bash,role=no-copy]
----
{"user_id": 2324, "event_type": "BUTTON_CLICK", "ts": "2024-11-25T20:23:59.380Z"}
----

=== Topic with schema (`value_schema_id_prefix` mode)

> In this example, it is assumed you have created the `ClickEvent` topic, set `redpanda.iceberg.mode` to `value_schema_id_prefix`, and are connecting to a REST-based Iceberg catalog. The following is an Avro schema for `ClickEvent`:

Reviewer: ```suggestion
Assume that you have created the `ClickEvent` topic, set `redpanda.iceberg.mode` to `value_schema_id_prefix`, and are connecting to a REST-based Iceberg catalog. The following is an Avro schema for `ClickEvent`:
```

[,avro]
----
{
    "type" : "record",
    "namespace" : "com.redpanda.examples.avro",
    "name" : "ClickEvent",
    "fields" : [
       { "name": "user_id", "type" : "int" },
       { "name": "event_type", "type" : "string" },
       { "name": "ts", "type": "string" }
    ]
 }
----

. Register the schema under the `ClickEvent-value` subject:
+
[,bash]
----
rpk registry schema create ClickEvent-value --schema path/to/schema.avsc --type avro
----

. Produce to the `ClickEvent` topic using the following format:
+
[,bash]
----
echo '"key1" {"user_id":2324,"event_type":"BUTTON_CLICK","ts":"2024-11-25T20:23:59.380Z"}' | rpk topic produce ClickEvent --format='%k %v\n' --schema-id=topic
----
+
The `value_schema_id_prefix` requires that you produce to a topic using the Schema Registry wire format, which includes the magic byte and schema ID in the prefix of the message payload. This allows Redpanda to identify the correct schema version in the Schema Registry for a record. See the https://www.redpanda.com/blog/schema-registry-kafka-streaming#how-does-serialization-work-with-schema-registry-in-kafka[Understanding Apache Kafka Schema Registry^] blog post to learn more about the wire format.

> 

Reviewer: The link goes to a subsection of the blog, but maybe the whole sentence should change. I searched the blog for "wire format" and "wire" and got nothing.

Reviewer 2: Updated - still mention wire format, but not explicitly linking it to the referenced blog post

+
[,sql]
----
SELECT *
FROM `<catalog-name>`.redpanda.ClickEvent;
----
+
[,bash,role=no-copy]
----
+-----------------------------------+---------+--------------+--------------------------+
| redpanda                          | user_id | event_type   | ts                       |
+-----------------------------------+---------+--------------+--------------------------+
| {"partition":0,"offset":0,"timestamp":2025-03-05 15:09:20.436,"headers":null,"key":null} | 2324    | BUTTON_CLICK | 2024-11-25T20:23:59.380Z |
+-----------------------------------+---------+--------------+--------------------------+
----

=== Topic in key-value mode

In `key_value` mode, you do not associate the topic with a schema in the Schema Registry, which means using semi-structured data in Iceberg. The record keys and values can have an arbitrary structure, so Redpanda stores them in https://apache.github.io/iceberg/spec/?h=spec#primitive-types[binary format^] in Iceberg.

> In this example, it is assumed you have created the `ClickEvent_key_value` topic, and set `redpanda.iceberg.mode` to `key_value`.

Reviewer: ```suggestion
In this example, assume that you have created the `ClickEvent_key_value` topic, and set `redpanda.iceberg.mode` to `key_value`.
```

| {"partition":0,"offset":0,"timestamp":2025-03-05 15:14:30.931,"headers":null,"key":key1} | {"user_id":2324,"event_type":"BUTTON_CLICK","ts":"2024-11-25T20:23:59.380Z"} |
+-----------------------------------+------------------------------------------------------------------------------+
----

Depending on your query engine, you might need to first decode the binary value to display the record key and value using a SQL helper function. For example, see the https://spark.apache.org/docs/latest/api/sql/index.html#unhex[`decode` and `unhex`^] Spark SQL functions, or the https://docs.snowflake.com/en/sql-reference/functions/hex_decode_string[HEX_DECODE_STRING^] Snowflake function. Some engines may also automatically decode the binary value for you. 

=== Time travel queries

Some query engines, such as Spark, support time travel with Iceberg, allowing you to query the table at a specific point in time. You can query the table as it existed at a specific timestamp or version number. 

Redpanda automatically removes expired snapshots on a periodic basis, which also reduces the window available for time travel queries.

> The following is an example of querying a `ClickEvent` table at a specific timestamp in Spark:

Reviewer: ```suggestion
The following example queries a `ClickEvent` table at a specific timestamp in Spark:
```
